### PR TITLE
Support dark mode

### PR DESCRIPTION
**Pull request by Fullstop000**
UCP https://github.com/pingcap-incubator/tidb-dashboard/issues/124

test

### Features
- Dark theme styles of  `antd` and `Fluent UI` will be automatically handled
- Use `*.module.dark.less` to specific dark theme styles in apps and components

### Some implementation details:
1. Group dark mode style files `*.dark.less` into different files without importing them explicitly  by webpack `entry` and specialized loader rules
2. Dynamically load the dark mode stylesheets by detecting the currently active app and the `manifest.json` generated by webpack

### Unresolved Problems

- Not support HMR
- In dev mode, `Files successfully emitted, waiting for typecheck results...` will be displayed after compiling and goes away in a short time. This does not influent the development though.
- ~`darkMode` global state implementation is buggy.~
- ~`HighlightSQL` is using `moize`, which is hard to using `CustomEvent` to update cached elements~

### CODE REVIEW

File: ui/lib/utils/themeSwitch.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
+import assetPrefix from '@dashboard/publicPathPrefix'
+import { fromEvent } from 'rxjs'
```
Comment (breeswish) [inline ui/lib/utils/themeSwitch.ts:2]: I think react's context is enough for this purpose.

File: ui/public/index.html [hunk 2]
```diff
@@ -1,19 +1,102 @@
...
-    <title>TiDB Dashboard</title>
-  </head>
-  <body>
-    <noscript>You need to enable JavaScript to run this app.</noscript>
-    <div id="dashboard_page_spinner"></div>
-    <div id="root"></div>
-  </body>
-</html>
+
+<head>
```
Comment (breeswish) [inline ui/public/index.html:4]: no need to change this file?

File: ui/dashboardApp/layout/main/Sider/Banner.module.dark.less [hunk 3]
```diff
@@ -0,0 +1,10 @@
+@import '~antd/es/style/themes/dark.less';
```
Comment (breeswish) [inline ui/dashboardApp/layout/main/Sider/Banner.module.dark.less:1]: what kind of styles needs a `.dark` variant?

File: ui/dashboardApp/layout/main/Sider/index.module.dark.less [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  overflow-x: hidden;
+  overflow-y: auto;
+  transition: none;
+  user-select: none;
+  background: linear-gradient(@sider-background, @component-background);
+  :global {
+    .ant-menu-submenu-selected {
+      color: #fff;
+    }
+    .ant-layout-sider-children {
```
Comment (breeswish) [inline ui/dashboardApp/layout/main/Sider/index.module.dark.less:20]: Why do we need to include non-color styles in `.dark` variant?

File: ui/config-overrides.js [hunk 5]
```diff
@@ -47,18 +51,175 @@ const enableEslintIgnore = () => (config) => {
...
+      new MiniCssExtractPlugin({
+        filename: 'static/css/[name].[contenthash:8].css',
+        ignoreOrder: true,
+      })
+    )(config)
   }
   return config
 }
 
+const addCustomLessLoader = (
```
Comment (breeswish) [inline ui/config-overrides.js:80]: why do we need these whole duplicated code?

File: ui/config-overrides.js [hunk 6]
```diff
@@ -47,18 +51,175 @@ const enableEslintIgnore = () => (config) => {
...
+            importLoaders: 2,
+            sourceMap: isEnvProduction && shouldUseSourceMap,
+          },
+          cssLoaderOptions,
+          customeCssModules
+        ),
+        'less-loader'
+      ),
+    },
+    {
```
Comment (breeswish) [inline ui/config-overrides.js:201]: Can we avoid duplication by cloning existing `use` rules?

File: ui/config-overrides.js [hunk 7]
```diff
@@ -157,8 +318,67 @@ const ignoreMiniCssExtractOrder = () => (config) => {
...
+        filename: 'static/css/[name].[contenthash:8].css',
+        ignoreOrder: true,
+      })
+      break
+    }
+  }
+  return config
+}
+
+const overrideManifestPlugin = () => (config) => {
```
Comment (breeswish) [inline ui/config-overrides.js:331]: There are too many duplicate codes here, please simply them..

File: ui/config-overrides.js [hunk 8]
```diff
@@ -157,8 +318,67 @@ const ignoreMiniCssExtractOrder = () => (config) => {
   for (let i = 0; i < config.plugins.length; i++) {
     const p = config.plugins[i]
     if (!!p.constructor && p.constructor.name === 'MiniCssExtractPlugin') {
-      const miniCssExtractOptions = { ...p.options, ignoreOrder: true }
-      config.plugins[i] = new MiniCssExtractPlugin(miniCssExtractOptions)
+      config.plugins[i] = new MiniCssExtractPlugin({
+        filename: 'static/css/[name].[contenthash:8].css',
```
Comment (breeswish) [inline ui/config-overrides.js:322]: please comment the purpose of expliciting specifying the filename.

File: ui/dashboardApp/index.js [hunk 9]
```diff
@@ -22,7 +22,38 @@ import AppInstanceProfiling from '@lib/apps/InstanceProfiling/index.meta'
 import AppClusterInfo from '@lib/apps/ClusterInfo/index.meta'
 import AppSlowQuery from '@lib/apps/SlowQuery/index.meta'
 
+import { darkmodeEnabled, switchDarkMode } from '@lib/utils/themeSwitch'
+import publicPathPrefix from './publicPathPrefix'
+const path = require('path')
+
+function loadJSON(filelocation, callback) {
```
Comment (breeswish) [inline ui/dashboardApp/index.js:29]: I think you can use `axios`

File: ui/dashboardApp/index.js [hunk 10]
```diff
@@ -22,7 +22,38 @@ import AppInstanceProfiling from '@lib/apps/InstanceProfiling/index.meta'
...
+        resolve()
+      }
+    }
+  })
+  xobj.send(null)
+  return p
+}
+
+async function initManifest() {
+  const location = path.join(publicPathPrefix, 'asset-manifest.json')
```
Comment (breeswish) [inline ui/dashboardApp/index.js:46]: I think you can simply embed the `asset-manifest.json` file in the HTML content. See https://github.com/facebook/create-react-app/blob/f5c3bdb65480f93b2d4a4c2f3214fc50753de434/packages/react-dev-utils/InlineChunkHtmlPlugin.js

File: ui/lib/utils/themeSwitch.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    // remove all dark styles
+    for (const link of links) {
+      if (isDarkmodeStyle(link)) {
+        head.removeChild(link)
+      }
+    }
+  }
+}
+
+export function loadAppDarkStyles(appID: string): void {
```
Comment (breeswish) [inline ui/lib/utils/themeSwitch.ts:47]: can we simply use `__webpack_require__` member functions to load file?

File: ui/lib/utils/themeSwitch.ts [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+function newCSSLink(href: string): HTMLLinkElement {
+  const link: HTMLLinkElement = document.createElement('link')
+  link.rel = 'stylesheet'
+  link.type = 'text/css'
+  link.href = href
+  link.setAttribute(THEME_KEY, THEME_DARKMODE)
+  return link
+}
+
+const persistDarkmodeKey = '@@tidb_dashboard_darkmode'
```
Comment (breeswish) [inline ui/lib/utils/themeSwitch.ts:65]: `darkmode` is enough

**Comment by breeswish** [comment]
> `HighlightSQL is using moize`

The solution can be providing a wrapper over the memorized version that receives context.

File: ui/dashboardApp/index.js [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import './wdyr'
+// import './wdyr'
```
Comment (breeswish) [inline ui/dashboardApp/index.js:1]: Needs revert.

File: ui/config-overrides.js [hunk 14]
```diff
@@ -191,52 +431,75 @@ const supportDynamicPublicPathPrefix = () => (config) => {
...
+module.exports = {
+  webpack: override(
+    fixBabelImports('import', {
+      libraryName: 'antd',
+      libraryDirectory: 'es',
+      style: true,
+    }),
+    ignoreMiniCssExtractOrder(),
+    addCustomLessLoader(
+      {
```
[the remark this answers is not in the document]
Comment (Fullstop000) [reply ui/config-overrides.js:427]: Actually they're different 🤣

File: ui/dashboardApp/layout/main/Sider/Banner.module.dark.less (hunk 3 above)
Comment (Fullstop000) [reply ui/dashboardApp/layout/main/Sider/Banner.module.dark.less:1]: The dark styles for app pages and components

**Comment by breeswish** [comment]
Keep open.

**Comment by breeswish** [comment]
Dark mode is not our P1 tasks, close for now. It will be picked up again later.

**Pull request closed without merging**